apps/emails-and-messages/src/modules/sendgrid/configuration/sendgrid-configuration.service.ts

```diff
@@ -0,0 +1,287 @@
+
+    if (args.initialData) {
+      this.configurationData = args.initialData;
+    }
+  }
+
+  /**
+   * Fetch configuration from Saleor API and cache it.
+   * If configuration is not found, create a new, empty one.
+   */
+  private async pullConfiguration() {
+    logger.debug("Fetch configuration from Saleor API");
+
+    const config = (await this.metadataConfigurator.getConfig()) || { configurations: [] };
+
+    this.configurationData = config;
+  }
+
+  /**
+   * Push configuration to Saleor API
+   */
+  private async pushConfiguration() {
+    logger.debug("Push configuration to Saleor API");
+
+    await this.metadataConfigurator.setConfig(this.configurationData!);
+  }
+
+  /**
+   * Returns configuration from cache or fetches it from Saleor API
+   */
+  async getConfigurationRoot() {
+    logger.debug("Get configuration root");
+
+    if (this.configurationData) {
+      logger.debug("Using cached configuration");
+      return this.configurationData;
+    }
+
+    // No cached data, fetch it from Saleor API
+    await this.pullConfiguration();
+
+    if (!this.configurationData) {
+      logger.warn("No configuration found in Saleor API");
+
+      throw new SendgridConfigurationServiceError("API returned no configuration", "CANT_FETCH");
+    }
+
+    return this.configurationData;
+  }
+
+  // Saves configuration to Saleor API and cache it
+  async setConfigurationRoot(config: SendgridConfig) {
+    logger.debug("Set configuration root");
+
+    this.configurationData = config;
+    await this.pushConfiguration();
+  }
+
+  /**
+   * Get configuration for given ID. Throws if not found.
+   */
+  async getConfiguration({ id }: { id: string }) {
+    logger.debug("Get configuration");
+    const configurationRoot = await this.getConfigurationRoot();
+
+    const configuration = configurationRoot.configurations.find((conf) => conf.id === id);
+
+    if (!configuration) {
+      throw new SendgridConfigurationServiceError(
+        "Configuration not found",
+        "CONFIGURATION_NOT_FOUND"
+      );
+    }
+
+    return configuration;
+  }
+
+  /**
+   * Get list of configurations. Optionally filter them.
+   */
+  async getConfigurations(filter?: FilterConfigurationsArgs) {
+    logger.debug("Get configurations");
+    const configurationRoot = await this.getConfigurationRoot();
+
+    let filtered = configurationRoot.configurations;
+
+    if (!filter) {
+      return filtered;
+    }
+
+    if (filter.ids?.length) {
+      filtered = filtered.filter((c) => filter?.ids?.includes(c.id));
+    }
+
+    if (filter.active !== undefined) {
+      filtered = filtered.filter((c) => c.active === filter.active);
+    }
+
+    if (filter.availableInChannel?.length) {
+      filtered = filtered.filter((c) =>
+        isAvailableInChannel({
+          channel: filter.availableInChannel!,
+          channelConfiguration: c.channels,
+        })
+      );
+    }
+
+    return filtered;
+  }
+
+  /**
+   * Create a new configuration
+   */
```

Comment:
I suggest not adding comments like this :D 

![image](https://github.com/saleor/apps/assets/9268745/f648ebb6-56c5-4d94-af1e-842c71f5a271)